### PR TITLE
Add support for SASL XOAUTH2 to SmtpTransport.

### LINKED ISSUE #16815: CakePHP with Mailer + Microsoft with SASL XOAUTH2
### Description

Hi,
in connection with the withdrawal of basic verification by Microsoft, I would like to suggest introducing the new verification service to the mailer. From what I was able to determine, the situation is similar in the case of Google.
Link to the authorization description in Microsoft:
[https://learn.microsoft.com/en-us/exchange/client-developer/legacy-protocols/how-to-authenticate-an-imap-pop-smtp-application-by-using-oauth](https://learn.microsoft.com/en-us/exchange/client-developer/legacy-protocols/how-to-authenticate-an-imap-pop-smtp-application-by-using-oauth)

Unfortunately, I did not find anything about the topic in your documentation.

### CakePHP Version

4.0

### PR DESCRIPTION
**Pull request by ADmad**
Closes #16815

### CODE REVIEW

File: src/Mailer/Transport/SmtpTransport.php [hunk 1]
```diff
@@ -394,6 +397,26 @@ protected function _authLogin(string $username, string $password): void
...
+     * @param string $username Username.
+     * @param string $token Token.
+     * @return void
+     * @see https://learn.microsoft.com/en-us/exchange/client-developer/legacy-protocols/how-to-authenticate-an-imap-pop-smtp-application-by-using-oauth#smtp-protocol-exchange
+     * @see https://developers.google.com/gmail/imap/xoauth2-protocol#smtp_protocol_exchange
+     */
+    protected function _authXoauth2(string $username, string $token): void
+    {
+        $authString = base64_encode(sprintf(
+            "user=%s\1auth=Bearer %s\1\1",
```
Comment (markstory) [inline src/Mailer/Transport/SmtpTransport.php:431]: I have to wonder what led to the choice to use `^A` as the separator :thinking:

**Review by markstory — APPROVED** [review]
Looks good to me.

File: src/Mailer/Transport/SmtpTransport.php (hunk 1 above)
Comment (ADmad) [reply src/Mailer/Transport/SmtpTransport.php:431]: Yeah, without some background info that does seems like a pretty strange character choice.

**Comment by ADmad** [comment]
@michalfordonski Would you like to try out this branch and confirm that it works for you as expected?

**Comment by michalfordonski** [comment]
@ADmad Sure, how can I invoke the new authorization method? Key in config/app.php?

**Comment by ADmad** [comment]
@michalfordonski Just follow the docs for using the [`Smtp` transport](https://book.cakephp.org/4/en/core-libraries/email.html#configuring-transports) and set the token for the `password` config key.

**Comment by michalfordonski** [comment]
@ADmad 
I don't know, or I understand well:
If I have already checked the method of logging in via SMTP, then after starting this branch, the only difference is that instead of a password in the configuration under its key, it should be a token, right? If it is, then something is not working. The token is valid, up-to-date, and the return from the server is:
SMTP server did not accept the password.

**Comment by ADmad** [comment]
Hmm.. Upon connection `SmtpTransport` parses the server response and picks the the 1st supported auth type. So it seems the server supports `PLAIN`/`LOGIN` auths too and it's trying to use one of those auth types with the token as a password which expectedly fails.

Seems we need to add an option to explicitly specify the auth type too. 

Just to check could you modify `SmtpTranport` directly in core and add `$this->authType = self::AUTH_XOAUTH2` at the very beginning of the `_auth()` method and see if things work.

**Comment by michalfordonski** [comment]
@ADmad Now the message has changed. Unfortunately, it doesn't say much, so I sent a request for details to the department that has access to the configuration of our Microsoft products.
The error reads:
SMTP Error: 535 5.7.3 Authentication unsuccessful [random_id.constant_value.PROD.OUTLOOK.COM]

where strings are used instead of random_id and constant_value

When I have more information, I will add them in the thread.

**Comment by ADmad** [comment]
In the meantime I have added the ability to specify the auth type to use instead of implicitly picking the 1st one supported by the server.

**Comment by michalfordonski** [comment]
@ADmad 
@ADmad Different configurations still return this or the associated error. In the specification from the administrator, I have indications that the Microsoft sending integration should use TLS 1.0 or later. Is it so?

In the configuration, I have the EmailTransport.default.tls key set to true

SMTP Error: 530 5.7.57 Client not authenticated to send mail. Error: 535 5.7.3 Authentication unsuccessful

SMTP Error: 535 5.7.3 Authentication unsuccessful

**Comment by ADmad** [comment]
`'tls' => true` would enable TLS for the connection before authentication is done and in case of failure you would get an exception with message `SMTP server did not accept the connection or trying to connect to non TLS SMTP server using TLS.`

**Comment by dereuromark** [comment]
What shall we do with this one?

**Comment by othercorey** [comment]
Shall we merge this as-is?

**Pull request merged**